synapse/handlers/message.py

```diff
@@ -738,8 +746,16 @@ async def create_and_send_nonmember_event(
                     assert event.internal_metadata.stream_ordering
                     return event, event.internal_metadata.stream_ordering
 
+            prev_events = None
+            if "prev_events" in event_dict:
+                prev_events = event_dict["prev_events"]
```

Comment:
This needs a comment and can also be replaced by:

```suggestion
            prev_events = event_dict.get("prev_events")
```

This makes me a bit uncomfortable though. I just audited all the calls to `create_and_send_nonmember_event` and they (mostly) seem safe. A questionable call site is the `ModuleApi` -- this gets passed to various pieces of 3rd party code and would potentially give them the same power as an AS. Maybe realistically this is OK, but I'll want opinions of others.

Reply:
Should we just add the same `requester.app_service` condition around this to make it more straightforward on what can happen?

Reply:
cc @clokep 

(outdated hidden diff comment)

Reply:
Continued in https://github.com/matrix-org/synapse/pull/9247#discussion_r575407222